File: examples/CocosCreatorTs/assets/scripts/proto/ReqTestPhp.ts
import { Packet } from '@mi/mod/Packet'
import { MsgTestPhp } from './MsgTestPhp'


export class ReqTestPhp
{
	private _u64: Long;
	private _strxx: string;
	private _msg_req: MsgTestPhp;
	private msg_opt_flag: number = 0;
	private _msg_opt: MsgTestPhp;
	private _msg_rep: MsgTestPhp[] = [];


	public Encode(): Packet {
		let packet: Packet = new Packet();
		packet.WriteUlong(this._u64);
		packet.WriteString(this._strxx);
		packet.WriteBuffer(this._msg_req.GetBuffer());
		packet.WriteByte(this.msg_opt_flag);
		if (this.msg_opt_flag == 1)
		{
			packet.WriteBuffer(this._msg_opt.GetBuffer());
		}
		let msg_rep_count: number = this._msg_rep.length;
		packet.WriteUshort(msg_rep_count);
		for (var i: number = 0; i < msg_rep_count; i++)
		{
			let xxx: MsgTestPhp = this._msg_rep[i];
			packet.WriteBuffer(xxx.GetBuffer());
		}
		packet.Encode(40060);
		return packet;
	}


	public get u64(): Long { return this._u64; }
	public set u64(value: Long) { this._u64 = value; }
	public get strxx(): string { return this._strxx; }
	public set strxx(value: string) { this._strxx = value; }
	public get msg_req(): MsgTestPhp { return this._msg_req; }
	public set msg_req(value: MsgTestPhp) { this._msg_req = value; }
	public get msg_opt(): MsgTestPhp { return this._msg_opt; }
	public set msg_opt(value: MsgTestPhp) { this.msg_opt_flag = 1; this._msg_opt = value; }
	public get msg_rep(): MsgTestPhp[] {return this._msg_rep; }
	public set msg_rep(value: MsgTestPhp[]) { this._msg_rep = value; }
}

File: data/data.code.lua.client/ReqTestPhp.lua
module("ReqTestPhp", package.seeall)


function new()
	local tab_obj = {}
	setmetatable(tab_obj, {__index = ReqTestPhp})
	tab_obj:init()
	return tab_obj
end

function init(self)
	self.u64 = 0
	self.strxx = ""
	self.msg_req = nil
	self.msg_opt = nil
	self.msg_rep = {}
end


function encode(self)
	local pack = packet.new()
	pack:writeU64(self.u64)
	pack:writeString(self.strxx)
	pack:writeBytes(self.msg_req:getBytes())
	if self.msg_opt ~= nil then
		pack:writeU8(1)
		pack:writeBytes(self.msg_opt:getBytes())
	else
		pack:writeU8(0)
	end
	local msg_rep_count = table.getn(self.msg_rep)
	pack:writeU16(msg_rep_count)
	for i = 1, msg_rep_count do
		pack:writeBytes(table.remove(self.msg_rep):getBytes())
	end
	return pack:encode(Msg.P_REQ_TEST_PHP)
end

return ReqTestPhp
